### ISSUE TITLE
[Reproduction] Clarification on the 200-person personality dataset and PIP protocol/results (Tables 3/6/10)

### ISSUE THREAD

**Issue by kindness314** [zjutangk/PTCD#1]
Hi, thanks for releasing the PTCD code and data.

I'm trying to reproduce the MEL+PIP personality detection results reported in the paper. The released pipeline is running on my side, but after checking the public data, implementation, and reported metrics in some detail, I still cannot match the published results.

The main issue is that I am not sure whether the currently released personality data and PIP code are exactly the versions/configurations used for Tables 3 and 6.

I've separated the questions below because several of them are independent. A reply by question number would be very helpful.

---

## 1. Size of the personality detection dataset

The paper states that the personality detection dataset contains **200 famous characters**, with more than 200 personality-related entries per character on average.

However, from the personality detection data available through the Google Drive linked in the README, I get:

- train: 31 persons
- test: 69 persons
- total: 100 persons

The README describes `by-product` as the downstream personality detection data, and I could not find another download link or a preprocessing step that produces additional personality data.

The released PIP training script also directly reads:

```bash
data/train.csv
data/test.csv
```

### Q1

Is the currently released 100-person personality dataset the complete dataset used for the paper?

If not, where can the remaining data used for the reported personality detection experiments be obtained?

---

## 2. Person-level train / validation / test split

If the paper's stated 200-person dataset and the released 69-person test set refer to the same experiment population, then there would be:

```text
200 - 69 = 131
```

non-test persons.

The released training file contains only 31 persons, so under that assumption, 100 persons are not accounted for by the currently released train/test files.

The released 31-person training set is also nearly perfectly balanced on all four MBTI dimensions:

```text
I/E: 15 / 16
S/N: 15 / 16
T/F: 15 / 16
P/J: 15 / 16
```

### Q2

What exact person-level split was used for the PTCD personality detection experiment?

In particular:

- how many persons were used for training?
- was there a validation set?
- how many persons were used for testing?
- were all 200 persons used in the reported experiment?

---

## 3. Relationship between the released 69-person test set and Table 3

The released 69-person test set has the following label counts:

```text
I/E: I=19, E=50
S/N: S=55, N=14
T/F: T=10, F=59
P/J: P=46, J=23
```

Using these class counts, I checked the integer confusion matrices that would be possible for a single 69-person evaluation.

For each dimension, there is exactly one single-run integer confusion matrix whose ACC and per-dimension Macro-F1 round to the values reported in Table 3:

| Dimension | Confusion matrix | ACC | Macro-F1 |
|---|---|---:|---:|
| I/E | TN=16, FP=3, FN=0, TP=50 | 95.6522% | 94.2580% |
| S/N | TN=52, FP=3, FN=1, TP=13 | 94.2029% | 91.4815% |
| T/F | TN=8, FP=2, FN=1, TP=58 | 95.6522% | 90.8448% |
| P/J | TN=46, FP=0, FN=4, TP=19 | 94.2029% | 93.1548% |

Using the metric convention implemented in the released `utils/tools.py`, the corresponding positive-label F1 values for E/N/F/J are:

```text
E: 97.087379
N: 86.666667
F: 97.478992
J: 90.476190
```

Their mean is:

```text
92.927307% -> 92.93%
```

which matches the reported average F1.

I understand that Table 3 reports averages over 3 runs, so I do **not** take the single-run integer reconstruction above as proof of the actual per-run predictions. However, the numerical compatibility makes the released 69-person test set a strong candidate for the evaluation set used in the paper.

### Q3

Can you confirm whether the released 69-person `test.csv` is the PTCD personality test set used for Tables 3/6?

---

## 4. Exact Accuracy = 77.20%

Table 6 reports:

```text
ACC       94.94
F1        92.93
Exact ACC 77.20
```

The released evaluation code defines Exact Accuracy as the fraction of persons for whom all four MBTI dimensions are predicted correctly.

If Table 6 used the same fixed 69-person test set, a single-run Exact Accuracy must therefore be of the form:

```text
k / 69
```

No integer `k/69` rounds to 77.20%.

Table 3 explicitly states that its reported results are averaged over 3 runs. If Table 6 used the same fixed 69-person split and an ordinary arithmetic mean of three Exact Accuracy values, then the average would be equivalent to:

```text
(k1 + k2 + k3) / 207
```

and no integer `k/207` rounds to 77.20% either.

The nearest value is:

```text
160 / 207 = 77.294686%
```

### Q4

How was Exact Accuracy = 77.20% in Table 6 calculated?

In particular:

- does Table 6 use the same three-run protocol as Table 3?
- were the same 69 test persons used in every run?
- what were the individual Exact Accuracy values?
- how were the runs aggregated?

---

## 5. Backbone and PIP training configuration

There also seems to be some ambiguity around the BERT backbone.

Section 6.1 states `bert-base-uncased`, while the released personality detection script uses:

```bash
--bert_model "bert-base-chinese"
```

and the paper later also refers to BERT-base-Chinese.

The released `personality_detection/scripts/run.sh` uses:

```text
train_batch_size = 32
num_pretrain_epochs = 50
```

### Q5

Which exact encoder checkpoint/tokenizer and PIP training configuration were used for Tables 3/6?

It would be helpful to know:

- BERT checkpoint
- tokenizer
- batch size
- number of epochs
- learning rate
- warmup
- optimizer
- random seeds

---

## 6. Scheduler calculation

In the released `train.py`, the scheduler step count is calculated with:

```python
int(len(train_data) / train_batch_size) * num_train_epochs
```

while the actual training loop runs for:

```python
num_pretrain_epochs
```

The default value of `num_train_epochs` is 34, while the released `run.sh` sets `num_pretrain_epochs=50`.

With the released 31-person training set and batch size 32:

```text
int(31 / 32) = 0
```

so the scheduler is initialized with zero planned optimization steps in my current Transformers environment.

### Q6

Was this scheduler calculation used in the experiment that produced the reported results, or was a different scheduler implementation used?

---

## 7. Padding in the Set Decoder

The released dataloader pads each person's utterance feature matrix to the maximum number of utterances in the batch.

As far as I can tell, the released PMA/MAB path does not receive an attention/padding mask, so padded zero vectors can participate in attention.

I tested the effect directly. Without masking, the prediction for a person can change when the other persons in the same batch change. After adding a padding mask, that batch-composition dependence disappears in my test.

I also compared masked and unmasked training pipelines using the same corrected scheduler. Both sets of selected checkpoints were then evaluated with batch size 1 to remove test-time padding:

```text
No mask:
ACC   66.18 ± 11.91
F1    68.86 ± 6.31
Exact 19.32 ± 25.94

With padding mask:
ACC   80.56 ± 0.75
F1    75.39 ± 4.36
Exact 52.17 ± 2.90
```

Because checkpoint selection followed each training/evaluation path, I treat this as evidence that padding handling materially affects the released pipeline, rather than as a pure causal estimate of the training-mask effect.

### Q7

Was a padding/attention mask used in the PIP experiment that produced the reported results?

If so, is the corresponding implementation available?

---

## 8. LayerNorm in MAB

The MAB formulation in the paper includes LayerNorm, while the released `Set_dec` uses:

```python
ln = False
```

I tested a version with the paper-style LayerNorm together with the corrected scheduler and padding masking.

Across 3 seeds, I obtained:

```text
ACC   82.73 ± 0.92
F1    73.16 ± 7.99
Exact 57.49 ± 1.67
```

This changes the results, but still does not approach the reported:

```text
ACC   94.94
F1    92.93
Exact 77.20
```

### Q8

Was LayerNorm enabled in the Set Decoder / MAB used for the reported personality detection results?

---

## 9. Checkpoint selection / validation protocol

The released `train.py` evaluates `test.csv` after every epoch and updates the best model using:

```python
if acc > self.best_acc:
```

where `acc` comes from the test-set evaluation.

I could not find a separate validation split in the released PIP pipeline.

### Q9

How was the final checkpoint used for the reported results selected?

Was it based on:

- validation performance,
- test performance,
- the final epoch,
- or another criterion?

If a validation set was used, was it part of the 200-person dataset?

---

## 10. PIP loss

On the released single-GPU path, training uses `loss_bce_ent`.

The entropy part contains:

```python
probs = torch.tensor(logits)
```

which detaches the logits, and the following entropy expression appears to apply logarithms directly to raw logits as if they were probabilities.

In my environment, this can result in `NaN` values in the reported training loss, although the BCE term still provides gradients.

The paper describes binary cross-entropy for personality detection.

### Q10

Which loss implementation was actually used for the reported PIP experiment?

Was it:

- BCE only,
- the released `loss_bce_ent`,
- or another implementation?

---

## 11. F1 value in Table 10

I also noticed that the reported F1 values are not consistent across the result tables:

```text
Table 3 / Table 6:
F1 = 92.93

Table 10:
F1 = 94.94
```

The ACC and Exact values otherwise match the corresponding reported result.

### Q11

Is `F1 = 94.94` in Table 10 a typo, or was Table 10 produced under a different evaluation setting?

---

## Current reproduction status

For reference, on the released 31-train / 69-test split, after:

- correcting the scheduler calculation,
- adding padding masking,
- enabling the LayerNorm described in the paper,

I obtain the following 3-seed result:

```text
ACC   82.73 ± 0.92
F1    73.16 ± 7.99
Exact 57.49 ± 1.67
```

The checkpoints here were selected following the test-ACC-based selection logic of the released training pipeline.

As an additional diagnostic, even when looking across all tested seeds/epochs and considering the best observed value for each metric separately, I only obtain:

```text
ACC   83.70
F1    78.35
Exact 59.42
```

These are diagnostic oracle values over the configurations I tested, not a normal model-selection result.

At this point, the remaining difference seems most likely to involve the complete personality data, the exact experiment configuration, or both.

For convenience, the questions are:

1. Is the released 100-person personality dataset complete, and where is the rest of the 200-person data if not?
2. What exact train/validation/test split was used?
3. Is the released 69-person `test.csv` the test set used for Tables 3/6?
4. How was Exact Accuracy = 77.20% in Table 6 calculated and aggregated?
5. Which BERT checkpoint/tokenizer and PIP training settings/seeds were used?
6. Which scheduler calculation was used in the actual experiment?
7. Was padding masking used?
8. Was LayerNorm enabled?
9. How was the final checkpoint selected?
10. Which PIP loss implementation was used?
11. Is the Table 10 F1 value of 94.94 a typo or a different evaluation result?

I'm not assuming that the reported results are incorrect. I'm mainly trying to identify which parts of the released data/code differ from the setup used to produce the published results, so that the experiment can be reproduced faithfully.

Thanks again for making the project public. Any clarification would be greatly appreciated.
